src/Cms/System/UpdateStatus.php

```diff
@@ -0,0 +1,708 @@
+	 * is available
+	 */
+	public function messages(): array|null
+	{
+		if (isset($this->messages) === true) {
+			return $this->messages;
+		}
+
+		if (
+			$this->data === null ||
+			$this->currentVersion === null ||
+			$this->currentVersion === ''
+		) {
+			return null;
+		}
+
+		$type = $this->pluginName ? 'plugin' : 'kirby';
+
+		// collect all matching custom messages
+		$filters = [
+			'kirby'  => $this->app->version(),
+			'php'    => phpversion()
+		];
+
+		if ($type === 'plugin') {
+			$filters['plugin'] = $this->currentVersion;
+		}
+
+		$messages = static::filterArrayByVersion(
+			$this->data['messages'] ?? [],
+			$filters
+		);
+
+		// add a message for each vulnerability
+		// the current version is affected by
+		foreach ($this->vulnerabilities() as $vulnerability) {
+			if ($type === 'plugin') {
+				$vulnerability['plugin'] = $this->pluginName;
+			}
+
+			$messages[] = [
+				'text' => I18n::template(
+					'system.issues.vulnerability.' . $type,
+					null,
+					$vulnerability
+				),
+				'link' => $vulnerability['link'] ?? null,
+				'icon' => 'bug'
+			];
+		}
+
+		// add special message for end-of-life versions
+		$versionEntry = $this->versionEntry();
+		if (($versionEntry['status'] ?? null) === 'end-of-life') {
+			$messages[] = [
+				'text' => match ($type) {
+					'plugin' => I18n::template(
+						'system.issues.eol.plugin',
+						null,
+						['plugin' => $this->package->name()]
+					),
+					default => I18n::translate('system.issues.eol.kirby')
+				},
+				'link' => $versionEntry['status-link'] ?? 'https://getkirby.com/security/end-of-life',
+				'icon' => 'bell'
+			];
+		}
+
+		return $this->messages = $messages;
+	}
+
+	/**
+	 * Returns the raw status value
+	 *
+	 * @return string 'up-to-date'|'not-vulnerable'|'security-update'|
+	 *                'security-upgrade'|'update'|'upgrade'|'unreleased'|'error'
+	 */
+	public function status(): string
+	{
+		return $this->targetData()['status'];
+	}
+
+	/**
+	 * Version that is suggested for the update/upgrade
+	 */
+	public function targetVersion(): string|null
+	{
+		return $this->targetData()['version'];
+	}
+
+	/**
+	 * Returns the Panel theme for the status value
+	 *
+	 * @return string 'positive'|'negative'|'info'|'notice'
+	 */
+	public function theme(): string
+	{
+		return match ($this->status()) {
+			'up-to-date', 'not-vulnerable' => 'positive',
+			'security-update', 'security-upgrade' => 'negative',
+			'update', 'upgrade' => 'info',
+			default => 'notice'
+		};
+	}
+
+	/**
+	 * Returns the most important human-readable
+	 * status information as array
+	 */
+	public function toArray(): array
+	{
+		return [
+			'icon'          => $this->icon(),
+			'label'         => $this->label(),
+			'name'          => $this->pluginName,
+			'theme'         => $this->theme(),
+			'url'           => $this->url(),
+			'version'       => $this->currentVersion ?? '?',
+			'targetVersion' => $this->targetVersion() ?? $this->currentVersion ?? '?'
```

Comment:
@bastianallgeier I guess you added this fallback to the current version to display the latest version even if the version is already up to date, right?

There is also the case of an unreleased version (current version is higher than the latest version), in this case the fallback will break. If we want to display the latest version, we can get it with `$this->data['latest'] ?? null` more reliably.

Suggestion: What if the dropdown displays three versions like this (now that we have the room):

<img src="https://user-images.githubusercontent.com/1595007/191598661-27631f6b-f838-4f12-89ee-cb92e6cf8696.png" alt="Screenshot" width="320">

The data can come from here:

```suggestion
			'currentVersion' => $this->currentVersion ?? '?',
			'latestVersion'  => $this->data['latest'] ?? '?',
			'targetVersion'  => $this->targetVersion()
```

`targetVersion` explicitly has *no* fallback. If it's null, the dropdown would hide that "Suggested version" line entirely and only display current and latest version.

Reply:
Thinking about it, the suggested version is never needed as it's already contained in the status label. So I think current version and latest version are enough:

```suggestion
			'currentVersion' => $this->currentVersion ?? '?',
			'latestVersion'  => $this->data['latest'] ?? '?',
```